Src/VimCore/Modes_Insert_InsertMode.fs

```diff
@@ -528,6 +534,37 @@ type internal InsertMode
         // edit data for the session
         _textChangeTracker.CompleteChange()
 
+        // If applicable, use the effective change for the combined edit.
+        if
+            not _globalSettings.AtomicInsert
+            && _textChangeTracker.IsEffectiveChangeInsert
+        then
+            match _textChangeTracker.EffectiveChange with
+            | Some span when span.Length <> 0 ->
+
+                // Override the piecewise combined edit command with
+                // the effective change which is a simple insertion.
+                span
+                |> SnapshotSpanUtil.GetText
+                |> TextChange.Insert
+                |> InsertCommand.OfTextChange
+                |> Some
+                |> x.ChangeCombinedEditCommand
+
+                // If the effective change is confined to one line then
+                // move the caret to the end of the effective change.
+                // This can happen if the editor auto-inserted matching
+                // parentheses, quotes, brackets, etc. and the user
+                // didn't "finish" them manually. We want the caret
+                // and the final edit point to appear as if the user
+                // had entered them directly.

```

Comment:
This part makes me a bit nervous. The IDE puts a lot of work into the caret position around auto-completion of parens and such. Moving the caret here can end up subverting expectations of users here.

Reply:
Mode-less editors like Visual Studio must necessarily assume the user might want to "add some more text". But VsVim has the luxury of the user giving us a definitive indication that insert mode is over by typing the escape key.

Let me pose a question. Without this change, if the user issues a change command and inserts `WriteLine("Hello, world!")` where the closing quote and closing parenthesis were auto-inserted, after typing escape the cursor would rest on the exclamation point. If that change command is then repeated, the cursor would be now positioned on the closing parenthesis, which seems counter-intuitive.

To put it another way, in vim, after any insertion that doesn't break the undo sequence, the cursor is *always* at the end of the text that got inserted.

But, to put this in context, I agree that the guiding principle should be *not* to subvert the user's expectations, if possible. It's hard to know how users will react, but I can see an argument for both points of view. One thing I am sure of is that when the change is repeated, the user *definitely* wants those auto-inserted characters to be repeated as if they had typed them.

Reply:
I reverted this as it is only a minor feature. Moving the caret in that case makes sense to *me*, but the important thing is the change that gets repeated.